File: python/p_0010.py
#!/usr/bin/env python3

"""
Euler problem 10:

The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.
"""


import argparse


def compute_primes_sum_sieve(limit):
    """
    limit: upper limits for the prime numbers to compute
    return: sum of the prime numbers found below the limit
    """
    marked = [0] * limit
    value = 3
    s = 2
    while value < limit:
        if marked[value] == 0:
            s += value
            i = value
            while i < limit:
                marked[i] = 1
                i += value
        value += 2
    return s


if __name__ == '__main__':
    """
    Find the sum of all the primes below limit.
    """
    parser = argparse.ArgumentParser()
    parser.add_argument("limit", help="Upper limit for the primes to sum", type=int)
    args = parser.parse_args()
    print("Find the sum of the primes below", args.limit)
    print("Sum:", compute_primes_sum_sieve(args.limit))

File: cpp/p_0010.cc
#include <cstdint>
#include <iostream>
#include <vector>

// Function to compute the sum of primes below a given limit using the Sieve of
// Eratosthenes
uint64_t euler_0010(uint64_t limit) {
    if (limit < 2) return 0;

    // false means prime, true means marked (not prime)
    std::vector<bool> marked(limit, false);
    uint64_t sum{2};  // Start with 2, the only even prime

    // Mark multiples of primes starting from 3
    for (uint64_t value{3}; value < limit; value += 2) {
        if (!marked[value]) {
            sum += value;
            // We start marking from the squared value as the smallest multiples
            // are already taken care of.
            for (uint64_t i{value * value}; i < limit; i += value) {
                marked[i] = true;
            }
        }
    }

    return sum;
}

int main() {
    auto result = euler_0010(2'000'000);
    std::cout
        << "Problem 0010: The sum of all the primes below two million is: "
        << result << '\n';

    return 0;
}
